Sources/YUL/AST.swift

```diff
@@ -0,0 +1,342 @@
+//
+//  AST.swift
+//  YUL
+//
+//  Created by Yicheng Luo on 11/16/18.
+//
+
+import Foundation
+
+public typealias Identifier = String
+
+public enum Literal: CustomStringConvertible {
+  case num(Int)
+  case string(String)
+  case bool(Bool)
+  case decimal(Int, Int)
+  case hex(String)
+
+  public var description: String {
+    switch self {
+    case .num(let i):
+      return String(i)
+    case .string(let s):
+      return "\"\(s)\""
+    case .bool(let b):
+      return b ? "1" : "0"
+    case .decimal(let (n1, n2)):
+      return "\(n1).\(n2)"
+    case .hex(let h):
+      return h
+    }
+  }
+}
+
+public struct Block: CustomStringConvertible {
+  public var statements: [Statement]
+
+  public init(_ statements: [Statement]) {
```

Comment:
I would prefer some variadic arguments for this, makes the call-sites much neater.

`public init(_ statements: [Statement])` => `public init(_ statements: Statement...)`

https://docs.swift.org/swift-book/LanguageGuide/Functions.html